lib/posix-futex/futex.c

```diff
@@ -101,50 +102,57 @@ static int futex_wait(uint32_t *uaddr, uint32_t val, const __nsec *timeout)
 	struct uk_thread *current = uk_thread_current();
 	struct uk_futex f = {.uaddr = uaddr, .thread = current};
 
-	if (ukarch_load_n(uaddr) == val) {
-		irqf = ukplat_lcpu_save_irqf();
-		uk_spin_lock(&futex_list_lock);
-
-		/* Enqueue thread to wait list */
-		uk_list_add_tail(&f.list_node, &futex_list);
+	if (ukarch_load_n(uaddr) != val) {
+		uk_pr_debug("FUTEX_WAIT: Condition =%"PRIu32" not met (uaddr: %p)\n",
+			    val, uaddr);
+		return -EAGAIN;
+	}
 
-		uk_spin_unlock(&futex_list_lock);
-		ukplat_lcpu_restore_irqf(irqf);
+	/* Futex word _does_ contain expected val */
+	uk_pr_debug("FUTEX_WAIT: Condition =%"PRIu32" met (uaddr: %p)\n",
+			val, uaddr);
 
-		if (timeout)
-			/* Block at most until `timeout` nanosecs */
-			uk_thread_block_until(current, (__snsec)*timeout);
-		else
-			/* Block indefinitely */
-			uk_thread_block(current);
+	/* Enqueue thread to wait list */
+	irqf = ukplat_lcpu_save_irqf();
+	uk_spin_lock(&futex_list_lock);
+	uk_list_add_tail(&f.list_node, &futex_list);
+	uk_spin_unlock(&futex_list_lock);
+	ukplat_lcpu_restore_irqf(irqf);
 
-		uk_sched_yield();
+	if (timeout) {
+		/* Block at most until `timeout` nanosecs */
+		uk_pr_debug("FUTEX_WAIT: Wait %"__PRIsnsec" nsec for wake-up\n",
+				(__snsec) (*timeout));
+		uk_thread_block_until(current, (__snsec) (*timeout));
```

Comment:
Minor: maybe turn the expression `(__snsec) (*timeout)` into a variable since it is used multiple times?

Reply:
With this patch, it is only used a second time in the debug message that I add. However, normally this message is not compiled in, so I find adding a variable that stores the dereferenced value of `timeout` for a single occurrence unnecessary.